specs/form-associated-custom-element.md

```diff
@@ -0,0 +1,114 @@
+  2. Decorators cannot (easily?) provide a *default* implementation that can be overridden by the extending class. Using an abstract class allows straight overrides of implementation or merging of implementation with super
+
+The implementation will standardize an interface between two common cases: browsers with form-associated-custom-element support and browsers without.
+
+### Browsers with FACE support
+For browsers *with* support, methods and properties will generally proxy to the `ElementInternals` implementation.
+
+### Browsers with FACE support
+The implementation will manage a "proxy" element that will be appended to the light-dom. This proxy element will server as the custom-element's association to the form. All relevant properties will be forwarded to this element, and certain API calls will retrieve data from it.
+
+### API
+#### IDL attributes
+- `public static formAssociated: boolean`
+  - Required for form-associated custom element API. This will feature-detect the API and resolve a value corresponding to feature-support
+- `public readonly validity: ValidityState`
+  - Returns the [validity](https://developer.mozilla.org/en-US/docs/Web/API/ValidityState) of the component
+- `public readonly form: HTMLFormElement | null`
+  - The associated form element
+- `public readonly validationMessage: string`
+  - The current validation message of the element
+- `public readonly labels: Node[]`
+  - Labels associated to the element. See [risks and challenges](#retrieving-labels)
+- `protected abstract proxy: HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement`
+  - A proxy element that will be appended to the DOM when form-association APIs are unavailable. It will server as the custom-elements connection to the form.
+- `protected elementInternals: ElementInternals`
+  - Provides form-association through dedicated APIs.
+- `protected setFormValue(value: File | string | FormData | null, state?: File | string | FormData): void`
+  - When using `elementInternals`, this will set the value in the form. With no form-association support, this will noop because the value will automatically be associated by the proxy element. This can be overridden as necessary by components with more advanced behavior.
+- `protected onLabelClick(): void`: **Discussion**
+  - see [risks and challenges](#clicking-labels)
+- `protected handleKeyPress(): void`
+  - Will submit the form when `enter` key is pressed to match standard input behavior. Can be overridden or extended.
+- `protected setValidity(flags: ValidityStateFlags, message?: string, anchor?: HTMLElement): void;`
+  - With form association support, will set the validity of the component. With no form association, this will noop unless a message is passed. If a message is passed, the proxy element's `setCustomValidity` method will be invoked with the value of the message.
```

Comment:
I see "form association support" and "form-association support", which one should we use in the document?

Reply:
I'm definitely interested in how the implementations might look for the above. If there's a prototype, I'd love to have you walk me through it.

Reply:
I've posted a link to the prototype branch in the PR description. Not all of this is represented yet but I'll continue to flesh that prototype out.

Reply:
Cleaned this up